geo/src/algorithm/line_split/line_split_result.rs

```diff
@@ -0,0 +1,92 @@
+/// The result of splitting a line using
+/// [LineSplit::line_split()](crate::algorithm::LineSplit::line_split) method.
+/// It can contain between one and two [Lines](crate::Line) / [LineStrings](crate::LineString).
+/// 
+/// Note that it may not be desireable to use a `match` statement directly on this type if you only
+/// ever want one part of the split. For this please see the helper functions;
+/// [.first()](LineSplitResult#method.first),
+/// [.second()](LineSplitResult#method.second),
+/// [.into_first()](LineSplitResult#method.into_first), and
+/// [.into_second()](LineSplitResult#method.into_second).
+/// 
+/// ```
+/// # use geo::{LineString, coord};
+/// # use geo::LineSplit;
+/// 
+/// # let my_line = LineString::from(vec![coord! {x: 0., y: 0.},coord! {x: 1., y: 0.},]);
+/// if let Some(result) = my_line.line_split_twice(0.2, 0.5) {
```

Comment:
Thinking about this a bit more... is there a reason to not just have `line_split_many` and return `Vec<Option<T>>`? The guarantee is that the vec has n+1 entries for n input splits, and any of them might be None. Are the enums easier to use in the offset work?

Reply:
Hi @dabreegster :) 

Regarding `line_split_twice` I think you may be right that `line_split_many` returning `Vec<Option<T>>` is more broadly useful. I personally have a specific use-case for `line_split_twice()` where I want to [chop out a road segment between two chainages](https://github.com/thehappycheese/megalinref/blob/main/readme.md#34-use-linestring_from_road_slk), but I can just move that function into my own project.

I think for a single `line_split` the enum (`LineSplitResult<T>`) is helpful as explained above.

Maybe your comment is because it comes wrapped in an option `line_split(..) -> Option<LineSplitResult<Self>>`  which is an annoying extra step to unwrap?

Just thinking out loud; the possible outcomes are
a)
```rust
None, // returned when the input line has NAN or INF values, or the input `fraction` is NAN
Some(LineSplitResult::First       (T   )),
Some(LineSplitResult::Second      (   T)),
Some(LineSplitResult::FirstSecond (T, T)),
```

The tuple version is
b)
```rust
None, // returned when the input line has NAN or INF values, or the input `fraction` is NAN
Some((Some(T), None   )),
Some((None,    Some(T))),
Some((Some(T), Some(T))),
```

One level of wrapping can be removed like this;
c)
```rust
(None,     None   ), // returned when the input line has NAN or INF values, or the input `fraction` is NAN
(Some(T), None   ),
(None,    Some(T)),
(Some(T), Some(T)),
```

So why not just use the last one?

---

The most satisfying API would be if we are allowed to assume the input geometry and `fraction` are well-formed, then the result could be like this:

d)
```rust
LineSplitResult::First       (T   ),
LineSplitResult::Second      (   T),
LineSplitResult::FirstSecond (T, T),
```

In this case, the enum is helpful, because it prevents the user from expecting `(None,     None   )` which can never happen if the inputs are valid. Therefore the enum stops rust from complaining about non-exhaustive match arms.

---

I think it depends on whether `line_split` should be guarding against NAN/INF. Possibly not?  Otherwise if other algorithms rely on this algorithm, and they all check for NAN every time... there will end up being a lot of NAN checks? On the other hand, if it is user facing, then it probably should guard against invalid input?

Apologies for the wall-of-text and thank you for your time :)

Reply:
Thanks for explaining further! I think guarding against invalid input is necessary, so some way to represent that sounds helpful. I'm just trying to figure out why the enum is helpful vs returning a bunch of `Option<T>`s.

> I personally have a specific use-case for line_split_twice() where I want to [chop out a road segment between two chainages](https://github.com/thehappycheese/megalinref/blob/main/readme.md#34-use-linestring_from_road_slk), but I can just move that function into my own project.

I can imagine use cases for different numbers of split points. What about if some of the variations returned an array of fixed size, like `line_split -> [Option<T>; 2]` and you used https://doc.rust-lang.org/reference/patterns.html#slice-patterns? If the input is invalid, `[None, None]`. Otherwise the 3 variations can be represented. I just tried out slice patterns and https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=596e6e8d97ae0e673428dff4334c0b3f seems pretty succinct, and would save you from defining the enums.

> In this case, the enum is helpful, because it prevents the user from expecting (None,     None   ) which can never happen if the inputs are valid. Therefore the enum stops rust from complaining about non-exhaustive match arms.

Is `(None, None) => unreachable!("input should be valid")` in user code something to avoid? Similar to an `line_split(...).expect("input should be valid")`, right?